### ISSUE TITLE
Benchmarking scripts for decision tree

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#886]
We have a new decision tree classifier in `src/mlpack/methods/decision_tree/`, but we don't know how it compares against other libraries.  So we should write some benchmarking scripts for it, for inclusion in the https://github.com/zoq/benchmarks repository.

The implementation there can be heavily based off existing benchmarking scripts, like for instance

https://github.com/zoq/benchmarks/blob/master/methods/mlpack/decision_stump.py

And then we should also benchmark against other libraries, like Weka, Shogun, and possibly MATLAB.  (A PR is already open for Shogun's CART implementation: https://github.com/zoq/benchmarks/pull/16 --- it could probably be finished.)

**Comment by sam0410** [comment]
I am interested in writing the benchmark for the tree classifier in MATLAB. Do I need to rewrite the methods in decision tree in MATLAB?

**Comment by sam0410** [comment]
Hi @rcurtin !
I have read the code in the decision tree and I am interested in writing the benchmark for the same in MATLAB.

**Comment by rcurtin** [comment]
Yes, if you'd like to write the decision tree MATLAB benchmark, that would be great.  You should take a look at the other MATLAB benchmarks in the benchmarks repository and imitate the format there.  The directions in that repository should be good enough that once you add the script and modify the configuration, you should be able to run the benchmarks and see the results.

**Comment by sam0410** [comment]
Added a pull request in [https://github.com/zoq/benchmarks](https://github.com/zoq/benchmarks) (decision_tree.m) for MATLAB benchmark

**Comment by sam0410** [comment]
Hi @rcurtin !
I tested the code using the following datasets in mlpack/tests/data using
TRAINx='iris_train.csv';
TRAINy='iris_train_labels.csv';
TESTx='iris_test.csv' ;
TESTy='iris_test_labels.csv';

and obtained the following results-
accuracy of prediction: 95.238095
total_time taken: 0.020836s

**Comment by IlievskiV** [comment]
Are there any benchmarks for Weka? I was working recently with the Weka source code, so I might figure it out how to make one.

**Comment by rcurtin** [comment]
Not for the decision tree, no---it would be easy to add though I think.

**Comment by Iron-Stark** [comment]
@rcurtin Any new developments on this issue?

I came here via the mailing list archives (wrt the better benchmarking project for gsoc). If no one is working on this, I'd love to try out this one.

Thanks!

**Comment by sam0410** [comment]
Hi @Iron-Stark !
I am almost done with the MATLAB benchmarking. You may take up any other :)

**Comment by Iron-Stark** [comment]
@sam0410  I'll start off with scikit learn.

**Comment by Iron-Stark** [comment]
@rcurtin @zoq 

I was going through the benchmark code for RandomForestClassifier and found that while parsing the input options using regex there were only four options (n_estimators, criterion, max_depth and seed) handled in the code. However, scikit-learn provides many other options like min_samples_split, min_samples_leaf, min_weight_fraction_leaf, max_features etc in their RandomForest Implementation. Is this intentional?

I was hoping to write the benchmark code for DecisionTreeClassifier. Should I only provide options for criterion, max_depth and random state or should my code cover all the options given in scikit-learn documentation?

Sorry for any unintended confusion.

**Comment by zoq** [comment]
If you like you can provide options for the parameters as well. I'm not sure that the mentioned parameters already existed at the time I implemented the benchmark script for the scikit method. In general, the intention is to compare methods with the same parameter set and to use reasonable values if one implementation does not provide the same option as another implementation.

**Comment by ThyrixYang** [comment]
I've done some work on benchmark of mlpack_decision model several days ago. I noticed the issue about categorical data, and it is needed for processing some dataset(their label start  from 1, not 0). Should I make a pr now with less dataset support, or wait for that pr get merged?

**Comment by zoq** [comment]
Your call, a PR with a limited dataset support is just fine.

**Comment by AbdealiJK** [comment]
Can't see this on the impact website. Is this live?
I can only see knn stats

**Comment by rcurtin** [comment]
No, unfortunately I don't think we've rebuilt any benchmarking results with the decision tree yet.  There is an interactive interface at http://www.mlpack.org/benchmark.html.  However the database being used there is quite old.  (We are working on getting something more up-to-date deployed, but it takes a while to run all the benchmarks.)

**Comment by rickllyxu** [comment]
Interesting! I'd like to start to build a benchmarking against Shogun for my first contributing. 
@Iron-Stark  I do not think you should use RandomForestClassifier in sklearn. RandomForestClassifier is an ensemble method based on decision trees.

**Comment by rickllyxu** [comment]
well, sorry. Just now I check the benchmark, finding that there exists `decision_tree.py` in `shogun` directory.

**Comment by rcurtin** [comment]
There is still no script for decision trees in mlpack though, so you are more than welcome to do that if you like.

**Comment by rickllyxu** [comment]
thanks for your reply, I 'd like to take it for a starter.

**Comment by rcurtin** [comment]
This can be closed once benchmarks#115 is merged.

**Comment by rcurtin** [comment]
Fixed in mlpack/benchmarks#115.  Thanks @rickllxyu!
